pyro/infer/mcmc/hmc.py

```diff
@@ -28,31 +29,49 @@ class HMC(TraceKernel):
     :param int num_steps: The number of discrete steps over which to simulate
         Hamiltonian dynamics. The state at the end of the trajectory is
         returned as the proposal.
+    :param dict transforms: Optional dictionary that specifies a transform
+        for a sample site with constrained support to unconstrained space. The
+        transform should be invertible, and implement `log_abs_det_jacobian`.
+        If not specified and the model has sites with constrained support,
+        automatic transformations will be applied, as specified in
+        :mod:`torch.distributions.constraint_registry`.
     """
 
-    def __init__(self, model, step_size=0.5, num_steps=3):
+    def __init__(self, model, step_size=0.5, num_steps=3, transforms=None):
         self.model = model
         self.step_size = step_size
         self.num_steps = num_steps
+        self.transforms = {} if transforms is None else transforms
+        self._automatic_transform_enabled = True if transforms is None else False
         self._reset()
         super(HMC, self).__init__()
 
     def _get_trace(self, z):
+        # Since the model is specified in the constrained space, transform the
+        # unconstrained R.V.s `z` to the constrained space.
+        z_constrained = z.copy()
+        for name, transform in self.transforms.items():
+            z_constrained[name] = transform.inv(z_constrained[name])
         z_trace = self._prototype_trace
-        for name, value in z.items():
+        for name, value in z_constrained.items():
             z_trace.nodes[name]['value'] = value
         trace_poutine = poutine.trace(poutine.replay(self.model, trace=z_trace))
         trace_poutine(*self._args, **self._kwargs)
-        return trace_poutine.trace
+        return trace_poutine.trace, z_constrained
```

Comment:
I think that it is better to move this calculation for z_constrained to `potential_energy` method for three reasons:
- Keep the old API of `_get_trace`
- Avoid unnecessarily calculate `z_constrained` in the last step of `sample`.
- We can safely change the name `unconstrained_potential_energy` to `_potential_energy` without misunderstanding.

Reply:
@neerajprad do you want to make this change in this PR, or should I merge now and wait for a follow-up PR?

Reply:
@fehiepsi - We will be storing unconstrained values in the traces, in that case. So we will need to run the inverse transforms before we return the trace. I am happy to make that change, it might make things a bit more explicit.

Reply:
@fritzo - It's a small change; let me address it right now itself.

Reply:
Look great now!